### PR TITLE
Support produce request backwards compatibility with old magic versions

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

**Comment by apurvam** [comment]
Is this ready for review?

**Comment by hachikuji** [comment]
@apurvam @guozhangwang This is ready for review

File: clients/src/main/java/org/apache/kafka/clients/KafkaClient.java [hunk 1]
```diff
@@ -136,6 +136,7 @@ ClientRequest newClientRequest(String nodeId, AbstractRequest.Builder<?> request
      * @param callback the callback to invoke when we get a response
      */
     ClientRequest newClientRequest(String nodeId, AbstractRequest.Builder<?> requestBuilder,
-                                          long createdTimeMs, boolean expectResponse,
-                                          RequestCompletionHandler callback);
+                                   long createdTimeMs, boolean expectResponse,
+                                   RequestCompletionHandler callback);
+
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/KafkaClient.java]: Is this intentional?

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java [hunk 2]
```diff
@@ -77,8 +79,7 @@
 
     /**
      * Create a new record accumulator
-     * 
-     * @param batchSize The size to use when allocating {@link MemoryRecords} instances
+     *  @param batchSize The size to use when allocating {@link MemoryRecords} instances
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java]: Extra space.

File: clients/src/main/java/org/apache/kafka/common/requests/ProduceRequest.java [hunk 3]
```diff
@@ -46,12 +46,17 @@
...
         private final int timeout;
         private final Map<TopicPartition, MemoryRecords> partitionRecords;
 
-        public Builder(short acks, int timeout, Map<TopicPartition, MemoryRecords> partitionRecords) {
-            super(ApiKeys.PRODUCE);
+        public Builder(byte magic,
+                       short acks,
+                       int timeout,
+                       Map<TopicPartition, MemoryRecords> partitionRecords) {
+            super(ApiKeys.PRODUCE, (short) (magic >= LogEntry.CURRENT_MAGIC_VALUE ? 3 : 2));
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/common/requests/ProduceRequest.java]: This piece of logic is a bit unclear to me, will discuss offline.

**Pull request merged**